Sources/core/objects/Gun.cpp

```diff
@@ -0,0 +1,96 @@
+#include "Gun.h"
+#include "Tile.h"
+
+Gun::Gun(size_t id) : Item(id)
+{
+    SetSprite("icons/guns.dmi");
+    ammunition_ = 1;
+    max_ammunition_ = 1;
+}
+
+bool Gun::UseAmmo()
+{
+    if (ammunition_ <= 0)
+    {
+        ammunition_ = 0;
+        return false;
+    }
+    --ammunition_;
+    return true;
+}
+
+void Gun::Shoot(VDir target)
+{
+    id_ptr_on<CubeTile> tile = GetOwner()->GetOwner();
```

Comment:
I think, it is a bad approach.
1. An user of the gun does not care about projectile type or some random stuff. It is reasonable to request from him only target location/bullet vector. An user should not handle projectile types and sounds, that's the gun work.
2. It is not really good to hold sounds & projectile type as variables, because they unlikely will be changed. It is kind of per-class variable. So no any manipulations with sounds are needed.

So I strongly suggest you to make the protected 
`ShootImpl(VDir target, const std::string& projectile_type, const std::string& sound)`. 
function in `Gun`.
So you will be able to make something like that:
```
void LaserGun::Shoot(VDir target)
{
    std::string sound;
    // Somehow set the sound with GetRand() help or anything
    ShootImpl(target, Laser::T_ITEM_S(), sound);
}
```
Moreover, you can just make virtual `GetProjectileType` and `GetSoundOnShot`, which will be called inside of `Shoot`, and then `ShootImpl` will not be needed - in order to create a new gun type it will be needed to override `GetProjectileType` and `GetSoundOnShot`, and not `Shoot` function. But this may be a little bit overcomplicated and not flexible enough, so just `ShootImpl` should work fine.